[src] Created testing framework to test Bazel repository rules outside of Bazel invocation using pure shell scripting

ZachiNachshon opened the pull request:
[src] Added an example git repository named `small-repo` to be used by git repository rules test-kit
[src] Created x4 basic use cases for testing the `git_cached_repository` rule

ZachiNachshon committed:
[src] Created testing framework to test Bazel repository rules outside of Bazel invocation using pure shell scripting
[src] Added an example git repository named "small-repo" to be used by git repository rules test-kit
[src] Created x4 basic use cases for testing the git_cached_repository rule

ZachiNachshon committed:
[fix] Avoid redundant Bazel clean on newly create workspace
[src] Create a dedicated git cached folder for every test case
[src] Moved git cached repo rule related actions from bazel.sh to test file

ZachiNachshon committed:
[src] GitHub workflow - temp commit

ZachiNachshon committed:
[src] Changed git path to /usr/bin/git to allow git interaction in CI tests pipeline

Commit:
[src] Introduced a makefile for running tests rather than invoking scripts manually
[src] Added support for running tests in a dockerized container
[src] Configured git globally with dummy user & email
[src] Running git actions on host should send output to TEST_log to catch potential error logs

or-shachar commented on testing/e2e/init.sh:
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+
+# Configure git
+git config --global user.email "wix_build_tools@wix.com"
```
Maybe an even more fake domain would be better :)

ZachiNachshon replied:
I'll change the git configs but the Wix brand is in the repo name..

or-shachar commented on GNUmakefile:
```diff
@@ -0,0 +1,21 @@
+default: help
+
+.PHONY: run-tests
```
😮 makefile?  you sure you want this in the stack?

ZachiNachshon replied:
Makefile lets you manage the scripts you use in an organized manner knowing the exec folder. I prefer using it over executing random scripts from different locations which can cause issues.

or-shachar commented on testing/e2e/repos.sh:
```diff
@@ -0,0 +1,69 @@
+
+  mkdir -p $REPOS_DIR
+
+  cp ./testing/testdata/small-repo.tar.gz $REPOS_DIR
```
14kb of data seems rather big... any chance to get something that would be < 5k ?

ZachiNachshon replied:
The `.git` index contains x5 commits as can be seen [in here](https://github.com/wix-playground/wix_build_tools/blob/aaf49d790ae0efc5c7a2f1c2e1442e6aa0a01b09/testing/e2e/repos.sh#L28) for the testing framework to interact with.

Why do you think 14kb is big? Is there a way to exclude files / folders from being fetched when relying on `wix_build_rules` as a source dependent?

or-shachar replied:
I wish... git smart fetcher... I wish github had better places for dev binary resources and prod code.

or-shachar commented on testing/e2e/init.sh:
```diff
@@ -0,0 +1,34 @@
+git config --global user.name "Wix"
+
+if [[ -z "${TEST_TMPDIR:-}" ]]; then
+  export TEST_TMPDIR="$(mktemp -d ${TMPDIR:-/tmp}/bazel-test.XXXXXXXX)"
```
what is the `.XXXXX` thing?

ZachiNachshon replied:
Randomization of a temp directory name when using `mktemp` to create temp dirs.
For example - `bazel-test.WfCrj8UA`

or-shachar commented on testing/e2e/init.sh:
```diff
@@ -0,0 +1,34 @@
+    log_info "Test environment was cleaned already !"
+  fi
+
+  delete_git_cached_directory
```
This function is called from `rules.sh` but it's never loaded here... does it make sense?

ZachiNachshon replied:
It is the same behavior as you'll probably never instantiate a random file from a testing framework, rather, you'll read the docs or find by yourself the entry point of the testing framework.

Remember that we're talking about bash scripts.. it's not something structured like Scala/Java, and because of that, there are assumption such as which file(s) to run, execution path etc.. which have to be enforced.

Regardless, I've sourced relevant sh files for all custom defined functions - it'll be easier to figure out the function origin 👍🏽

or-shachar commented on testing/e2e/init.sh:
```diff
@@ -0,0 +1,34 @@
+  # Clean TEST_TMPDIR and verify path contains an identifier before deletion
+  if [[ -d ${TEST_TMPDIR} && "${TEST_TMPDIR}" == *"bazel-test"* ]]; then
+    rm -rf ${TEST_TMPDIR}
+    log_info "Cleaned local test environment"
```
log_info is from `logging.sh`... again - it's not loaded here ...

ZachiNachshon replied:
Sourced relevant sh files for all custom defined functions

or-shachar commented on testing/e2e/bazel.sh:
```diff
@@ -0,0 +1,38 @@
+#!/bin/bash
+
```
would be nice to have something that describes what this file is for.

I understand that each `*.sh` file gives you several new "bash functions" - but share with me (and with future you) what is the purpose of each file, If I were to look for a function - how do I know where to find it, if I were to add a new function - where should I add it?

Is it the bazel driver?

ZachiNachshon replied:
Sure, I'll add a description section on top of each file used by the testing framework 👍🏽

or-shachar commented on testing/e2e/bazel.sh:
```diff
@@ -0,0 +1,38 @@
+#!/bin/bash
+
+function write_workspace_file() {
+  cat > "$1" << EOF
```
I hate bash arguments... `$1` is so mysterious. 
I prefer that each function would have first lines naming each argument so I'll know that $1 is filepath and $2 is workspace name

ZachiNachshon replied:
Sounds legit, I'll add local vars, don't hate 😸

or-shachar commented on testing/e2e/bazel.sh:
```diff
@@ -0,0 +1,38 @@
+}
+
+function create_workspace_with_default_repos() {
+  write_workspace_file "${1:-WORKSPACE}" "${2:-wix_build_tools_tests}"
```
default value for filepath is simply `WORKSPACE`, not absolute? relative? Is it okay? (haven't read the usage yet, but I'm flagging here that it might be problematic).

ZachiNachshon replied:
The assumption is that every test starts on the `WORKSPACE` root, I tend to agree with you on that since it might be problematic (and redundant) to trace working paths in case of an issue.

I'll change it to absolute path.

or-shachar commented on testing/e2e/bazel.sh:
```diff
@@ -0,0 +1,38 @@
+
+EOF
+
+  echo $new_workspace_dir
```
Do we need this echo ? If we're leaving this print at least let's add something more useful like `"[Info] new workspace created at $location"`

or-shachar replied:
oh it's the return value... I see (hate bash functions, did I say that already?)

or-shachar commented on testing/e2e/rules.sh:
```diff
@@ -0,0 +1,41 @@
+}
+
+function copy_git_cached_repo_rules_to_workspace() {
+  if [[ -z "${RULES_GIT_DIR:-}" ]]; then
```
I prefer that we'll actually use rules from `wix_build_tools` and not copy single bzl files into the new workspace.
you can use `local_repository` on the WORKSPACE with relative/absolute path to your workspace

ZachiNachshon replied:
Excellent idea ! as part of the test env initialization, I'm copying the entire repository once to the test root folder. 
Every test dedicated `WORKSPACE` file is referencing the `wix_build_tools` as `local_repository` 👍🏽

or-shachar commented on testing/e2e/rules.sh:
```diff
@@ -0,0 +1,41 @@
+EOF
+}
+
+function delete_git_cached_directory() {
```
I didn't expect this one to be under `rules.sh`... but maybe I need to get used to it

ZachiNachshon replied:
The cached git directory is relevant to the rule `git_cached_repository` - `rules.sh` is indeed too general, will renaming the file to `git_rules_utils.sh` will make more sense?

or-shachar commented on testing/e2e/rules.sh:
```diff
@@ -0,0 +1,41 @@
+  fi
+
+  if [[ -d "${HOME}/.git-cache" ]]; then
+    rm -rf "${HOME}/.git-cache"
```
would be great if the test wouldn't touch my HOME dir but work with dedicated dirs only

ZachiNachshon replied:
The `git_cache_repository` rule uses a default directory at the home path for storing git indexes if no custom dir supplied. It is a behavior that must get tested so we can't overlook.
Since the testing suite should run within a temporary container, why does it bothers you?

or-shachar commented on testing/e2e/utils.sh:
```diff
@@ -0,0 +1,49 @@
+COLOR_GREEN='\033[0;32m'
+COLOR_WHITE='\033[1;37m'
+COLOR_NONE='\033[0m'
+
```
didn't review this file.... but same rules apply:
- If you use a function from other file - I'd expect to see it's `source` in the header.
- Make sure you're not overwriting existing bash functions / names
- Make sure that test files are local to test and not writing / reading stuff form global folders.

ZachiNachshon replied:
> If you use a function from other file - I'd expect to see it's source in the header.

Sourced relevant sh files for all custom defined functions.

> Make sure you're not overwriting existing bash functions / names

WDYM? noticed any potential area of it to happen?
If any case like this one could slip, it'll get caught by the testing suite, isn't it?

> Make sure that test files are local to test and not writing / reading stuff form global folders.

Tests aren't sharing workspaces (they use random names), custom git cached folder of every tests is unique to every test - both gets cleared upon tests completion.
Where did you notice shared usage of global folders?

or-shachar replied:
> WDYM? noticed any potential area of it to happen?

Didn't notice anything... didn't read through the file... but words like `write` or `read` or `log`  might be taken in bash.

> Where did you notice shared usage of global folders?

Only on tests that used `$HOME/.git-cache`

or-shachar commented on testing/e2e/testkit.sh:
```diff
@@ -0,0 +1,84 @@
+  echo "correct dorothy, we are in bazel world" >${REPO_SMALL_DIR}/quotes/the_wizard_of_oz.txt
+  /usr/bin/git add "quotes/the_wizard_of_oz.txt" >>${TEST_log}
+  /usr/bin/git commit -m "Updated a quote to generate a new commit" >>${TEST_log}
+  new_commit=$(/usr/bin/git log -n 1 --pretty=format:%H)
```
seems like you made a driver to create new commits on remote repository.
Why not generating this repo from scratch and drop usage in binary `tgz` file?

ZachiNachshon replied:
> Why not generating this repo from scratch and drop usage in binary tgz file?
I didn't quite understand your suggestion, you opt in for creating a new repository index on the fly appending a commit to it?

If this is the case, doesn't tests data should be immutable? this is why the initial repository (and its index) are added as a tgz file.

By exposing the way the repository-in-test is created and/or what its data contains, you potentially can break all the testing suites.

or-shachar commented on testing/e2e/testkit.sh:
```diff
@@ -0,0 +1,84 @@
+
+  if [[ ${cache_dir} == "default-dir" ]]; then
+    cat >${WORKSPACE_DIR}/WORKSPACE <<EOF
+load("//rules/git:git_cached_repository.bzl", "git_cached_repository")
```
see my comment about using `wix_build_tools` directly and not copying the rules into this repo.

ZachiNachshon replied:
Done, it is now references a `local_repository` i.e.

```
load("@wix_build_tools//rules/git:git_cached_repository.bzl", "git_cached_repository")
```

or-shachar commented on testing/e2e/testkit.sh:
```diff
@@ -0,0 +1,84 @@
+  local cache_dir=${2:-"default-dir"}
+
+  # Trim consecutive // characters since Bazel would fail if such exists on a target path
+  local trimmed_path=$(echo ${WORKSPACE_DIR} | tr -s '/')
```
who's populating WORKSPACE_DIR? is it a github actions thing? beware of coupling with the build server solution...

ZachiNachshon replied:
The testing framework exports the `WORKSPACE_DIR` as appears [in here](https://github.com/wix-playground/wix_build_tools/blob/aaf49d790ae0efc5c7a2f1c2e1442e6aa0a01b09/testing/e2e/bazel.sh#L27), it relies on the temporary folder created for the testing suite.

or-shachar commented on testing/git_cached_repository/git_cached_repository.sh:
```diff
@@ -0,0 +1,124 @@
+ROOT=$PWD
+
+function before_test() {
+  set_up
```
idea - if `set_up` is loaded from `unittest.sh` then let's call it `unittest_set_up` - it'll help me in understanding the flow better.

ZachiNachshon replied:
Renamed.

or-shachar commented on testing/git_cached_repository/git_cached_repository.sh:
```diff
@@ -0,0 +1,124 @@
+  before_test "test_cached_dir_created_as_expected"
+
+  # Given I declare a git cache rule with custom cache directory
+  create_git_cached_rule_for_small_repo "014459e6361b66a7b210758d4bf93f3a46ca5e88" ${TEST_GIT_CACHE_DIR}
```
what is this `014459e6361b66a7b210758d4bf93f3a46ca5e88`? random?

ZachiNachshon replied:
It is the commit-hash used for the `commit` attribute of `git_cached_repository` rule.
I've created a var named `commit_hash` though it seems a bit verbose, specifically when inlining bash function calls

or-shachar commented on testing/git_cached_repository/git_cached_repository.sh:
```diff
@@ -0,0 +1,124 @@
+  create_git_cached_rule_for_small_repo "014459e6361b66a7b210758d4bf93f3a46ca5e88" ${TEST_GIT_CACHE_DIR}
+
+  # And I create a shell library to reference the git cache rule
+  target_label=$(create_sh_lib_ref_small_repo)
```
target label should include the `//` part , no?

ZachiNachshon replied:
Correct, I've appended the `//` within `create_sh_lib_ref_small_repo` function rather than on the tests.

or-shachar commented on testing/git_cached_repository/git_cached_repository.sh:
```diff
@@ -0,0 +1,124 @@
+
+  # And I build the shell library target label
+  bazel build //${target_label} >&${TEST_log} ||
+    echo "Expected git cached directory to get created successfully"
```
why do we `||` here? Maybe we just don't want the bazel command to fail the build? Have you considered using `set -x` and `set +x` ?

ZachiNachshon replied:
We don't, the failure logs would get appended to `TEST_log`. Post failure, you'll have the Bazel log printed for the failing test and a clean message at the bottom of the test as follows:

```bash
---
...
Bazel logs..
...

Assertion error: Expected git cached directory to get created successfully
---
```

`set -x` and `set +x` are the opposite of each other, I might missed your suggestion on where to use them..
Appending logs to `TEST_log` and printing it on test failure in an organized section is better suitable imo of printing every bash command using `set -x` or hiding via `set +x` - it is using globals when you want fine grained logging without the actual bash noise.

or-shachar replied:
Oh I see :) cool! darn... it'll take me a while before I understand our way around here

or-shachar commented on testing/git_cached_repository/git_cached_repository.sh:
```diff
@@ -0,0 +1,124 @@
+    echo "Expected git cached default directory to get created successfully"
+
+  # Then I expect the repository to be cached successfully within default cache library
+  expect_folder "$HOME/.git-cache"
```
I'd avoid this test... unless you're running the test in a sandbox with dedicated HOME dir... it's intrusive to the developer's HOME dir.

ZachiNachshon replied:
Testing suites run within a container, locally and CI.

or-shachar commented on testing/git_cached_repository/git_cached_repository.sh:
```diff
@@ -0,0 +1,124 @@
+    echo "Expected to successfully fetch repo by commit hash"
+
+  # Then I expect the binary response to return the expected static string value
+  expect_log "toto, I have a feeling we are not in maven anymore"
```
🤣

or-shachar reviewed (changes requested):
Zachi :) 
1. Loved the Wizard of Oz reference. Very cute 😄 
2. Critical thing is that I think the tests shouldn't touch HOME directory of the developers (unless it's a dedicated HOME folder per test).
3. Important thing is that you use functions from files without calling `source` on them... you rely on the main test file to `source` all the files, but it makes it less readable.
4. I made a suggestion on how functions can be named so we can know that they come from your testing framework and from which file exactly.
5. I'm against checking in a full git repo. If you already made a mechanism for committing changes, it's only few more steps to generate a git repository without pushing binaries here.

Confession - Reading this - I wonder if we made the right tradeoff here when we chose bash testing over python... I have a feeling that python would have made it A LOT more readable and maintainable, and short. I'm guessing this can be rewritten in the future but I'm afraid that if we start rolling with this direction - it'll stay stuck like this for a while.

Please see my comments 🌹

or-shachar commented on testing/e2e/init.sh:
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+
+# Configure git
+git config --global user.email "wix_build_tools@wix.com"
```
Also - what will happen if I run the tests locally - will it change my local git user email and name? Any way to do it using local env variable that will be set by the test?

ZachiNachshon replied:
This is actually why it is advised to run within a container 😄 
I'll add a validation / read from sessioned git env vars to avoid tampering with local environment if testing suite executed locally

ZachiNachshon replied:
Done.

ZachiNachshon committed:
[src] Prevent copying specific rules to test env - rather copy the entire wix_build_tools repo and reference via local_repository rule
[src] Added description on top tests fw files and renamed a few
[src] Renamed test execution run-tests to run-tests-ci
[src] Source sh files from every other file they are being used
[fix] Do not configure git client when running on non containerized env if it is already configured

ZachiNachshon commented:
> Loved the Wizard of Oz reference. Very cute 😄

Thanks :blush:

> Critical thing is that I think the tests shouldn't touch HOME directory of the developers (unless it's a dedicated HOME folder per test).

Tests are running in a containerized environment when running locally. On CI they are running "locally" since they are wrapped by GitHub actions containerized environment.

> Important thing is that you use functions from files without calling source on them... you rely on the main test file to source all the files, but it makes it less readable.

Sourced sh files from every other file they are being used

> I made a suggestion on how functions can be named so we can know that they come from your testing framework and from which file exactly.

Couldn't find your suggestion, used the GitHub suggestion feature?

> I'm against checking in a full git repo. If you already made a mechanism for committing changes, it's only few more steps to generate a git repository without pushing binaries here.

This is exactly the way Bazel is testing the git repo rules, this testing framework is inspired by theirs and I think that it is better to use an immutable repo as I've replied [to this comment](https://github.com/wix-playground/wix_build_tools/pull/4#discussion_r621068306).

> Confession

It depends on the use case, for this one I think it will be easier to use straight forward bash commands rather than using a python subprocess for different shell script/command that you require to execute or creating wrappers for python libs so short it won't be necessarily

In addition as we've discussed previously, python has its environmental issues that we have and still witnessing on a daily basis that shouldn't be a problem with straight forward bash scripting.

And by looking at the Bazel shell testing framework, seems like it is good to be close to how to source operates :smile:

Let's give it a run since this was our initial choice which is already implemented and see how it goes, specially when we could use it to check other Bazel rules that we have internally.

ZachiNachshon committed:
[fix] Do not ignore exit codes

ZachiNachshon committed:
[src] Fixed failing test for git_cache_repo - should overwrite WORKSPACE file when creating a new commit hash to re-declare git_cache_repo rule
[src] Minor function rename

or-shachar commented on GNUmakefile:
```diff
@@ -0,0 +1,21 @@
+default: help
+
+.PHONY: run-tests-ci
```
Are we really sure we want to bring in GNUmakefile into the stack?

or-shachar commented:
> Couldn't find your suggestion, used the GitHub suggestion feature?

I don't know why it disappeared... :-( 
I suggested a new way for naming bash functions you declare on.

If the function comes from `assert.sh` then all public functions needs to be called `"assert_<name>"`

or-shachar commented on testing/e2e/bazel.sh:
```diff
@@ -0,0 +1,50 @@
+  local file_name=$1
+  local workspace_name=$2
+
+  export WIX_BUILD_TOOLS_DIR=${TEST_TMPDIR}/wix_build_tools
```
not sure you need to `export` it...  it may be good enough as local env var
Also - seems like this function assumes that the `wix_build_tools` is under `${TEST_TMPDIR}/wix_build_tools` 
Maybe a `local wix_build_tools_dir=$3` would be better?

ZachiNachshon replied:
You are correct, the export is wrongly used in here.. I'll export the `wix_build_tools` path on the function which actually creates that folder and use the exported var in here.

or-shachar commented on testing/e2e/bazel.sh:
```diff
@@ -0,0 +1,50 @@
+  write_workspace_file "${1:-WORKSPACE}" "${2:-wix_build_tools_tests}"
+}
+
+function create_new_random_workspace() {
```
Not sure I understand what this function creates?
WORKSPACE needs to have a WORKSPACE file :) but this method only creates random dir and adds in BUILD file

ZachiNachshon replied:
Correct, I'll rename to `create_new_workspace_random_directory`

ZachiNachshon replied:
Excluded the `BUILD` file creation to a separate method as well, no need to share responsibilities on a single function.

or-shachar commented on testing/e2e/git_rules_testkit.sh:
```diff
@@ -0,0 +1,98 @@
+
+function recreate_git_cached_rule_for_small_repo() {
+  # Re-create WORKSPACE file and re-declare git_cached_repository with the new commit hash
+  create_workspace_file_targeting_wix_build_tools "${WORKSPACE_DIR}/WORKSPACE"
```
Again - what is `WORKSPACE_DIR` and how does it gets assigned?

ZachiNachshon replied:
You might have missed the [previous comment](https://github.com/wix-playground/wix_build_tools/pull/4#discussion_r621070505) regarding the `WORKSPACE_DIR`.

> what is WORKSPACE_DIR

It is an exported variable created by the testing framework assigned with the temporary workspace directory path per test use case.

> how does it gets assigned

It is being created and assigned for every test case as can be seen [in here](https://github.com/wix-playground/wix_build_tools/blob/aaf49d790ae0efc5c7a2f1c2e1442e6aa0a01b09/testing/e2e/bazel.sh#L27)

or-shachar commented on testing/e2e/git_rules_testkit.sh:
```diff
@@ -0,0 +1,98 @@
+    remote_url = "${REPO_SMALL_DIR}",
+    commit = "${commit_hash}",
+    branch = "master",
+    cache_directory = "${cache_dir}",
```
Maybe there's a way for you to DRY here? 
Write the HEADER
```python
load("@wix_build_tools//rules/git:git_cached_repository.bzl", "git_cached_repository")
git_cached_repository(
    name = "small_repo",
    remote_url = "${REPO_SMALL_DIR}",
    commit = "${commit_hash}",
    branch = "master",
```

Then using `if` statememnt write/not-write the line:
```python
    cache_directory = "${cache_dir}",
```
then write the closing?

ZachiNachshon replied:
What about the next attribute that we would like to have like `branch`, `shallow_since` etc..?
I'll see how to create a dynamic behavior of accepting different function arguments and then assigning the relevant attribute to the `git_cached_repository` similar to that you've suggested.

or-shachar commented on testing/e2e/repos_loader.sh:
```diff
@@ -0,0 +1,75 @@
+
+  cwd=$(pwd)
+  cd $REPOS_DIR
+  tar zxf small-repo.tar.gz
```
I still vote for constructing the repo on the fly rather than tar.gz...

ZachiNachshon replied:
Creating repos on the fly in addition to the concerns I've raised before would force the framework to support dynamic commit hashes. 
IMO we need to add more test use cases and evolve the framework accordingly avoiding premature optimizations, we're talking about `14Kb`..

or-shachar reviewed (commented):
> Tests are running in a containerized environment when running locally. On CI they are running "locally" since they are wrapped by GitHub actions containerized environment.

What promises that when running locally they will run in "containerized env"? The makefile?

> This is exactly the way Bazel is testing the git repo rules, this testing framework is inspired by theirs and I think that it is better to use an immutable repo as I've replied to this comment.

I wonder if we can at least download this tar.gz from some other place and not push it to the repository. Google repositories usually have binaries (see all the jars in `bazel` and in the `intellij` plugins but in a git eco-system - this becomes problematic.

> It depends on the use case, for this one I think it will be easier to use straight forward bash commands rather than using a python subprocess for different shell script/command that you require to execute or creating wrappers for python libs so short it won't be necessarily
> 
> In addition as we've discussed previously, python has its environmental issues that we have and still witnessing on a daily basis that shouldn't be a problem with straight forward bash scripting.
> 
> And by looking at the Bazel shell testing framework, seems like it is good to be close to how to source operates 😄
> 
> Let's give it a run since this was our initial choice which is already implemented and see how it goes, specially when we could use it to check other Bazel rules that we have internally.

Very well then 😀 

Please see my last comments 🌹 Feel free to merge if you feel like it's ready :)

ZachiNachshon commented:
Reply to https://github.com/wix-playground/wix_build_tools/pull/4#discussion_r622838256

> Didn't notice anything... didn't read through the file... but words like write or read or log might be taken in bash.

I've renamed the function names that could potentially be taken already and used your `<filename>_<function-name>` suggestion.

> Only on tests that used $HOME/.git-cache

Running tests in a containerized environment shouldn't be an issue, this is a git_cached rule behavior that must be tested imo.

ZachiNachshon commented:
> > Couldn't find your suggestion, used the GitHub suggestion feature?
> 
> I don't know why it disappeared... :-(
> I suggested a new way for naming bash functions you declare on.
> 
> If the function comes from `assert.sh` then all public functions needs to be called `"assert_<name>"`

I've took your suggestion and used it on files such as `utils` and `assert` containing function names that might collide with others.

ZachiNachshon commented:
> What promises that when running locally they will run in "containerized env"? The makefile?

I want to believe that people don't tent to execute scripts from all over the place on repositories they aren't familiar with 😄 
makefile is self explanatory about the action it run with help description on each and the execution confined to the makefile relative location.

> I wonder if we can at least download this tar.gz from some other place and not push it to the repository. Google repositories usually have binaries (see all the jars in `bazel` and in the `intellij` plugins but in a git eco-system - this becomes problematic.

Please see premature optimization comment [in here](https://github.com/wix-playground/wix_build_tools/pull/4#discussion_r622900479), we'll surely get to it on the next stages of this framework. We'll probably find additional gaps to solve after test running it.

> Very well then 😀

🚀

ZachiNachshon committed:
[src] Simplified git cached rules creation method to have dynamic named-args-like behaviour
[src] Added prefix to function names which might collide
[src] Renamed a few functions name to be more self explanatory
[src] Removed redundant filegroup from git rules BUILD file
